### PR TITLE
Handle escaping of XPath string literals as best we can

### PR DESCRIPTION
**Pull request by danni**

### CODE REVIEW

**Comment by koterpillar** [comment]
So there's a function which uses string interpolation to return a string that can be percent-formatted, which is then used in one case to construct a string for interpolation. I am not sure what to think of that yet.

**Comment by danni** [comment]
At least it's cleaner than it was!

File: lettuce_webdriver/webdriver.py [hunk 1]
```diff
@@ -82,42 +86,45 @@ def goto(step, url):
...
-        (link_url, link_text)))
+        '//a[@href="%s"][contains(., %s)]' %
+        (link_url, string_literal(link_text))))
 
 
-# General
-@step('The element with id of "(.*?)" contains "(.*?)"$')
+@step('The element with id of "([^"]*)" contains "([^"]*)"$')
+@step("The element with id of '([^']*)' contains '([^']*)'$")
 def element_contains(step, element_id, value):
```
Comment (koterpillar) [inline lettuce_webdriver/webdriver.py]: Two more?

File: lettuce_webdriver/tests/html_pages/basic_page.html [hunk 2]
```diff
@@ -54,6 +54,7 @@
         <input type="submit" name="submit2" value="ส่งฟอร์ม" />
     </form>
     <a href="http://google.com/">Google</a>
+    <a href="http://google.com/">Welcome "User"</a>
```
Comment (koterpillar) [inline lettuce_webdriver/tests/html_pages/basic_page.html]: I can't help hearing this in Rinzler's voice.

File: lettuce_webdriver/webdriver.py [hunk 3]
```diff
@@ -82,42 +86,45 @@ def goto(step, url):
...
-                    '//a[@href="%s"][./text()="%s"]' %
-                    (link_url, link_text))))
+                    '//a[@href="%s"][./text()=%s]' %
+                    (link_url, string_literal(link_text)))))
 
 
-@step('I should see a link that contains the text "(.*?)" '
-      'and the url "(.*?)"$')
+@step('I should see a link that contains the text "([^"]*)" and the url "([^"]*)"$')
+@step("I should see a link that contains the text '([^']*)' and the url '([^']*)'$")
```
Comment (koterpillar) [inline lettuce_webdriver/webdriver.py]: Do we need `"` in URLs? If yes, there have to be 4 `step` calls. If no, 2 but still needs fixing.

File: lettuce_webdriver/webdriver.py [hunk 4]
```diff
@@ -82,42 +86,45 @@ def goto(step, url):
...
-        '//a[@href="%s"][contains(., "%s")]' %
-        (link_url, link_text)))
+        '//a[@href="%s"][contains(., %s)]' %
+        (link_url, string_literal(link_text))))
 
 
-# General
-@step('The element with id of "(.*?)" contains "(.*?)"$')
+@step('The element with id of "([^"]*)" contains "([^"]*)"$')
+@step("The element with id of '([^']*)' contains '([^']*)'$")
```
Comment (koterpillar) [inline lettuce_webdriver/webdriver.py]: Most likely 4.

File: lettuce_webdriver/webdriver.py [hunk 5]
```diff
@@ -82,42 +86,45 @@ def goto(step, url):
...
 def element_contains(step, element_id, value):
     return world.browser.find_element_by_xpath(str(
         'id("{id}")[contains(., "{value}")]'.format(
             id=element_id, value=value)))
 
 
-@step('The element with id of "(.*?)" does not contain "(.*?)"$')
+@step('The element with id of "([^"]*)" does not contain "([^"]*)"$')
+@step("The element with id of '([^']*)' does not contain '([^']*)'$")
```
Comment (koterpillar) [inline lettuce_webdriver/webdriver.py]: And so on...

File: lettuce_webdriver/webdriver.py [hunk 6]
```diff
@@ -381,13 +391,15 @@ def select_multi_items(step, select_name):
             select.select_by_visible_text(option)
 
 
-@step('The "(.*?)" option from "(.*?)" should be selected$')
+@step('The "([^"]*)" option from "([^"]*)" should be selected$')
+@step("The '([^']*)' option from '([^']*)' should be selected$")
```
Comment (koterpillar) [inline lettuce_webdriver/webdriver.py]: 4

File: lettuce_webdriver/util.py [hunk 7]
```diff
@@ -22,6 +22,20 @@
 # pylint:disable=invalid-name
 
 
+def string_literal(content):
+    """Choose a string literal that can wrap our string"""
+
+    if '"' in content and "'" in content:
+        # there is no way to escape string literal characters in XPath
+        raise ValueError("Cannot represent this string in XPath")
```
Comment (koterpillar) [inline lettuce_webdriver/util.py]: "Cannot represent string with both single and double quotes in XPath"?
"Cannot represent {{str}} in XPath"?

**Pull request closed without merging**